File: aws_interface/sdk/java/source/src/test/java/AuthTest.java
import org.junit.jupiter.api.*;
import java.util.concurrent.Callable;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.awaitility.Awaitility.*;

public class AuthTest {
	static AWSInterface client = new AWSInterface("https://e3t7xfb981.execute-api.ap-northeast-2.amazonaws.com/prod_aws_interface/ipxsNKvdpXUgzAMqg2JgdM");
	String test_email = new String("ttest@gmail.com");
	String test_password = new String("ttestpassword");
	String new_email = new String("testwt123@gmail.com");
	String new_password = new String("test1451password");
	static String user_id = "jK5iK6UgYnTdaRhow5yvAT";
	static String guest_id = null;
	String fb_token = "EAAhNpAi3ZAqkBAKQGKSDG0wFZCgrd4LjAZCqGu0ZAqY8chIeishU2GFvmOAD6ZCTjLSg1NEPhtxmNqrp5Cb9BuamBVJEWzhnC5RcZAWwkr49Tgy8CjGwUEvEZBaZCniuS5L9fhZCSuscLATjvZAwRxdPuNTfxLOnE7MLaqr4TqBWaZA8QVpxumefK5U6xLZAtRN1QbXAVJZATAtz7zXZBPyrReru4nO4F7SkBlMaS5qIsfhmsc4AZDZD";
	String start_key = new String("0");
	boolean received = false;
	
	//Declare Callable object for check if the response is received;
	private Callable<Boolean> responseReceived(){
		return new Callable<Boolean>() {
			public Boolean call() throws Exception{
				return received;
			}
		};
	}
	
	// Before each testcase, initialize received flag with false to use awaitility for async testing
	@BeforeEach
	public void initialiize() {
		received = false;
	}
	
	//authRegiste is commented since it register same ID;
	/*
	//Test authRegister
	@Test
	public void authRegisterTest() {
		client.authRegister(new_email, new_password, (response, hasError)->{
            assertEquals(response.get("item").getAsJsonObject().get("email").getAsString(), new_email);
            received = true;	
	        });
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}
	 */
	
	// Test authGetUser
	@Test
	public void authGetUserTest() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authGetUser(user_id, (response, hasError)->{
				assertEquals(response.get("item").getAsJsonObject().get("email").getAsString(), test_email);
				assertEquals(response.get("item").getAsJsonObject().get("id").getAsString(), user_id);
				received = true;
			});
		});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}

	@Test
	public void authLogoutTest() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authLogout((response, hasError)->{
	            assertNotNull(response.get("success"));
	            received = true;
	        });
		});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}

	//Test loginFacebook
	@Test
	public void authLoginFacebookTest() {
		client.authLoginFacebook(fb_token, (response, hasError)->{
			assertNotNull(response.get("session_id"));
	        received = true;
		});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}


	//Test authGetUsers with start_key
	@Test
	public void authGetUsers1Test() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authGetUsers(start_key, (response, hasError)->{
				assertTrue(response.get("items").getAsJsonArray().isJsonArray());
	            received = true;
	        });
		});
		//wait until responseReceieved turns into true;
		await().until(responseReceived());
	}
	
	//Test authGetUsers without start_key
	@Test
	public void authGetUsers2Test() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authGetUsers((response, hasError)->{
				assertTrue(response.get("items").getAsJsonArray().isJsonArray());
	            received = true;
	            });
			});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}
	
	//test AuthGuest with guest_id
	@Test
	public void authGuest1Test() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authGuest(guest_id,(response, hasError)->{
	            assertNotNull(response.get("session_id"));
	            received = true;
	            });
			});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}
	
	@Test
	//test AuthGuest without guest_id
	public void authGuest2Test() {
		client.authLogin(test_email, test_password, (rr, hE)->{
			client.authGuest((response, hasError)->{
	            assertNotNull(response.get("session_id"));
	            received = true;
	            });
			});
		//wait until responseReceived turns into true;
		await().until(responseReceived());
	}
	
}

File: aws_interface/sdk/nodejs/test/authTest.js
// authTest.js

const Client = require('../aws_interface.js');
const assert = require("assert");

var client = new Client.Client();
const new_email = "test21@email.com";
const new_password = "test21password";
const test_email = "ttest@gmail.com";
const test_password = "ttestpassword";
const fb_token = "";
const user_id = "jK5iK6UgYnTdaRhow5yvAT";

describe('Auth Test', function() {
    //Set timeout of 7 seconds for async testing.
    this.timeout(7000);

    //Register new user with [new_email], [new_password]
    //Function is commented since it gets already registered user when repeated.
    /*
    it('test authRegister', function(){
         client.authRegister(new_email, new_password, {})
             .then(function(body) {
                 assert.equal(body['item']['email'] == test_email);
             });
     });
    */

    // Login and logout
    it('test login and logout', function(done){
        client.authLogin(test_email, test_password)
            .then(function(){
                client.authLogout().then(function(response) {
                        assert.ok(response['success']);
                        done();
                    });
            });
    });


    //Login with facebook access token [fb_token]
    it('test authLoginFacebook', function(done){
        client.authLoginFacebook(fb_token)
            .then(function(body){
                assert.ok(body['session_id']);
                done();
            });
        });
    });


    //Get user information with [user_id]
    it('test authGetUser', function(done){
        //Extend timeout to 10 seconds
        this.timeout(10000);
        client.authLogin(test_email, test_password)
            .then(function(body) {
                client.authGetUser(user_id)
                    .then(function(response){
                        assert.equal(response['item']['email'], test_email);
                        assert.equal(response['item']['id'], user_id);
                        done();
                    });
            });
    });

    //Get a list of users
    it('test authGetUsers', function(done){
        client.authLogin(test_email, test_password)
            .then(function(body) {
                client.authGetUsers(null,)
                    .then(function(response) {
                        assert.ok(response['items']);
                        done();
                    });
            });
    });

    //Guest login
    it('test authGuest', function(done){
            client.authGuest(null)
                .then(function(response) {
                    assert.ok(response['session_id']);
                    done();
                });
    });